### PR TITLE
Include chunked phase2 script

### PR DESCRIPTION
**Pull request by emmorais**
Ongoing work. Currently getting the following error: 
```
Running `/media/eduardo/DATA/eduardo/Aleo/code/aleo-setup/target/release/setup2 --curve-type bw6 --chunk-size 131072 --batch-size 131072 --contribution-mode full --proving-system groth16 new --challenge-fname challenge --challenge-hash-fname challenge.verified.hash --phase1-fname ../../phase1-tests/phase1 --phase1-powers 18 --num-validators 1 --num-epochs 1`
/media/eduardo/DATA/eduardo/Aleo/code/aleo-setup/target/release/setup2: unrecognized option `--curve-type`
```

### CODE REVIEW

File: setup2/scripts/phase2_chunked.sh [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+echo $SEED2 > seed2
+
+function check_hash() {
+  test "`xxd -p -c 64 $1.hash`" = "`b2sum $1 | awk '{print $1}'`"
+}
+
+cargo $CARGO_VER build --release --bin setup2
+
+phase2_new="cargo run --release --features cli  -- new --curve-type $CURVE --chunk-size $CHUNK_SIZE --batch-size $BATCH --contribution-mode full"
+phase2_chunked="cargo run --release --bin setup2 --features cli  -- --curve-type $CURVE --chunk-size $CHUNK_SIZE --batch-size $BATCH --contribution-mode full --proving-system $PROVING_SYSTEM"
```
Comment (Pratyush) [inline setup2/scripts/phase2_chunked.sh:25]: This should come with an argument for the binary, right? That is, one of `new`, `contribute`, etc?

File: setup2/scripts/phase2_chunked.sh [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+phase2_1="cargo run --release --bin setup2 --features cli  -- --curve-type $CURVE --batch-size $BATCH --contribution-mode chunked --chunk-size $CHUNK_SIZE --seed seed1 --proving-system $PROVING_SYSTEM"
+phase2_2="cargo run --release --bin setup2 --features cli  -- --curve-type $CURVE --batch-size $BATCH --contribution-mode chunked --chunk-size $CHUNK_SIZE --seed seed2 --proving-system $PROVING_SYSTEM"
```
Comment (Pratyush) [inline setup2/scripts/phase2_chunked.sh:27]: Same for these two?

File: setup2/scripts/phase2_chunked.sh (hunk 1 above)
Comment (niklaslong) [reply setup2/scripts/phase2_chunked.sh:25]: Looks like we're still missing the `seed` and the `proving_system` in the related struct.

Comment (niklaslong) [reply setup2/scripts/phase2_chunked.sh:25]: The command is actually specified lower in the script.

File: setup2/scripts/phase2_chunked.sh (hunk 2 above)
Comment (niklaslong) [reply setup2/scripts/phase2_chunked.sh:27]: ditto

File: phase1-cli/src/bin/prepare_phase2.rs [hunk 3]
```diff
@@ -69,8 +69,8 @@ fn prepare_phase2<E: Engine + Sync>(opts: &PreparePhase2Opts) -> Result<()> {
     // Deserialize the accumulator
     let current_accumulator = Phase1::deserialize(
         &response_readable_map,
-        UseCompression::Yes,
-        CheckForCorrectness::Full,
+        UseCompression::No,
+        CheckForCorrectness::No,
```
Comment (Pratyush) [inline phase1-cli/src/bin/prepare_phase2.rs:73]: if we've already checked the output of Phase1, then we don't need to check for correctness, which makes the preparation step faster. This should be re-enabled if we did not check that the output of Phase 1 is in the appropriate prime-order group.

File: phase2/src/polynomial.rs [hunk 4]
```diff
@@ -90,20 +91,39 @@ fn dot_product_vec<C: AffineCurve>(
...
-        .reduce(|| C::Projective::zero(), |a, b| a + b)
+                (coeff, ind)
+            })
+            .collect::<Vec<_>>();
+        input.sort_unstable_by(|a, b| a.1.cmp(&b.1));
+        let input = input
+            .into_par_iter()
+            .map(|(coeff, _)| coeff.to_repr())
+            .collect::<Vec<_>>();
+        snarkvm_algorithms::msm::variable_base::VariableBaseMSM::multi_scalar_mul(coeffs, &input)
```
Comment (Pratyush) [inline phase2/src/polynomial.rs:110]: The dot product can be computed quickly using a variable base MSM. We do this if `input.len()` is large enough.

File: setup-utils/src/helpers.rs [hunk 5]
```diff
@@ -48,8 +54,9 @@ pub fn print_hash(hash: &[u8]) {
 
 /// Multiply a large number of points by a scalar
 pub fn batch_mul<C: AffineCurve>(bases: &mut [C], coeff: &C::ScalarField) -> Result<()> {
+    let coeff = coeff.to_repr();
     let mut points: Vec<_> = cfg_iter!(bases)
-        .map(|base| base.into_projective().mul(*coeff))
+        .map(|base| base.mul_bits(BitIteratorBE::new(coeff)))
```
Comment (Pratyush) [inline setup-utils/src/helpers.rs:59]: This change ensures that we use mixed addition, which makes the multiplication slightly faster.

File: setup1-contributor/src/utils.rs [hunk 6]
```diff
@@ -88,16 +88,15 @@ pub fn create_parameters_for_chunk<E: PairingEngine>(
...
     method: &str,
     path: &str,
     rng: &mut R,
 ) -> Result<String> {
-    let view_key = ViewKey::try_from(private_key)?;
     let address = Address::try_from(private_key)?.to_string();
 
     let message = format!("{} {}", method.to_lowercase(), path.to_lowercase());
-    let signature = hex::encode(&view_key.sign(message.as_bytes(), rng)?.to_bytes_le()?);
+    let signature = hex::encode(&private_key.sign(message.as_bytes(), rng)?.to_bytes_le()?);
```
Comment (Pratyush) [inline setup1-contributor/src/utils.rs:99]: Q: is this change warranted? Not sure if this breaks things (did the address scheme change since the code was written?)

File: setup1-contributor/src/utils.rs [hunk 7]
```diff
@@ -116,8 +115,8 @@ pub fn sign_contribution_state<R: Rng + CryptoRng>(
     let contribution_state = ContributionState::new(challenge_hash, response_hash, next_challenge_hash);
     let message = contribution_state.signature_message()?;
 
-    let view_key = ViewKey::<Testnet2Parameters>::from_str(signing_key)?;
-    let signature = hex::encode(&view_key.sign(message.as_bytes(), rng)?.to_bytes_le()?);
+    let private_key = PrivateKey::<Testnet2>::from_str(signing_key)?;
+    let signature = hex::encode(&private_key.sign(message.as_bytes(), rng)?.to_bytes_le()?);
```
Comment (Pratyush) [inline setup1-contributor/src/utils.rs:119]: Ditto

File: setup2/src/cli/new.rs [hunk 8]
```diff
@@ -56,77 +82,76 @@ pub struct NewOpts {
...
+    )]
+    pub new_challenge_fname: String,
+    #[options(help = "phase 1 file name", default = "phase1")]
+    pub phase1_fname: String,
+    #[options(help = "phase 1 powers")]
+    pub phase1_powers: usize,
+    #[options(help = "number of validators")]
+    pub num_validators: usize,
+    #[options(help = "number of epochs")]
+    pub num_epochs: usize,
```
Comment (Pratyush) [inline setup2/src/cli/new.rs:119]: I copied this from the Celo repo; happy to remove it if it ends up being unnecessary.

**Comment by Pratyush** [comment]
Update: in the latest commit, the error that Eduardo encountered is no longer present, so the `new` subcommand runs to completion. The remaining work involves implementing the rest of the commands specified in the script.

File: setup1-contributor/src/utils.rs (hunk 6 above)
Comment (Pratyush) [reply setup1-contributor/src/utils.rs:99]: cc @howardwu

**Comment by Pratyush** [comment]
In the latest commit, most tests are passing, except some tests in `setup1-verifier`. These tests are failing due to an inability to deserialize a `PrivateKey` from the hardcoded `ViewKey` bytes. We can no longer use the `ViewKey` because in the latest `snarkvm`, we use `PrivateKey`s to sign (whereas earlier we signed with `ViewKey`s). 

To get the tests passing, I can sample a new `PrivateKey` and hardcode this in the tests, but overall I'm not sure if the new behaviour (i.e. switching away from `ViewKey` signing) is wanted.

**Comment by niklaslong** [comment]
This may need a rebase after #448 is merged.

**Comment by Pratyush** [comment]
Merged `phase2` into this.
